File: Problem3_Search2DMatrix.py
# Time Complexity: O(m + n)
# Space Complexity: O(1)
# Did this code successfully run on Leetcode : Yes

# Solution:
"""
1. Consider 2 pointers for row and column and start the search from the element in last row and first column.
2. Since all rows and columns are sorted in ascending order, decrease the row or increase the column to move towards target.
3. Return true if the target is found, else return false.
"""

def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
    if not matrix or len(matrix) is 0:
        return False

    rows = len(matrix)
    cols = len(matrix[0])

    # Start searching from last row and first column element
    row = rows - 1
    col = 0

    # Decrease the row or increase the col to move towards the target
    while row >= 0 and col < cols:
        if target == matrix[row][col]:
            return True
        elif target < matrix[row][col]:
            row -= 1
        else:
            col += 1

    return False

File: Problem3_Search2DMatrix.cs
// Time Complexity : O(m + n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        int m = matrix.Length - 1;
        int n = matrix[0].Length - 1;
        int low = 0;
        int high = n;
        // Begin search from top right or bottom left corner, as we can have firm choice of which side to search
        // Starting from Top Right corner
        while(low <= m && high >= 0) {
            if(matrix[low][high] == target) {
                return true;
            }
            else if(matrix[low][high] < target) { // Search down
                low++;
            }
            else { // Search left
                high--;
            }
        }
    }
}